test/main_test.cpp

```diff
@@ -0,0 +1,218 @@
+      frameBefore.resources = {
+        {11, {12, 13, 14, 15, 16, 17, 18}},
+        {21, {22, 23, 24, 25, 26, 27, 28}}};
+      frameBefore.actions = {
+        {100, {{{111, 112}, 113, 114}, {{121, 122}, 123, 124}}},
+        {200, {{{211, 212}, 213, 214}, {{221, 222}, 223, 224}}}};
+      frameBefore.units = {
+        {1000, {{}, {}}},
+        {5000, {{}, {}}}};
+      Unit& u1000 = frameBefore.units[1000][0];
+      Unit& u2000 = frameBefore.units[1000][1];
+      Unit& u5000 = frameBefore.units[5000][0];
+      Unit& u6000 = frameBefore.units[5000][1];
+      auto propertyId = 1;
+      #define U(propertyName) \
+        u6000.propertyName = 1000 + ( \
+        u5000.propertyName = 3000 + ( \
+        u2000.propertyName = 1000 + ( \
+        u1000.propertyName = 1000 + propertyId++)));
+      U(id) U(x) U(y) U(health) U(max_health) U(shield) U(max_shield)
+      U(energy) U(maxCD) U(groundCD) U(airCD) U(flags) U(type)
+      U(armor) U(shieldArmor) U(size) U(pixel_x) U(pixel_y)
+      U(pixel_size_x) U(pixel_size_y) U(groundATK) U(airATK)
+      U(groundDmgType) U(airDmgType) U(groundRange) U(airRange)
+      U(velocityX) U(velocityY)
+      U(playerId) U(resources) U(buildTechUpgradeType)
+      U(remainingBuildTrainTime) U(remainingUpgradeResearchTime) U(spellCD)
+      U(associatedUnit) U(associatedCount)
+      u5000.visible = u1000.visible = true;
+      u6000.visible = u2000.visible = false;
+      u1000.orders = {{ 10, 11, 12, 13, 14 }, { -10, -11, -12, -13, -14 }};
+      u2000.orders = {{ 20, 21, 22, 23, 24 }, { -20, -21, -22, -23, -24 }};
+      u5000.orders = {{ 50, 51, 52, 53, 54 }, { -50, -51, -52, -53, -54 }};
+      u6000.orders = {{ 60, 61, 62, 63, 64 }, { -60, -61, -62, -63, -64 }};
+      u1000.command = { 101, 102, 103, 104, 105, 106 };
+      u2000.command = { 201, 202, 203, 204, 205, 206 };
+      u5000.command = { 501, 502, 503, 504, 505, 506 };
+      u6000.command = { 601, 602, 603, 604, 605, 606 };
+  
+      // Our Frame is complete.
+      // Let's see how it handles serialization!
+      std::stringstream buffer;
+      buffer << frameBefore;
+      buffer >> frameAfter;
+
+      // Convert unsorted maps to sorted vectors for comparison
+      auto sortedResourcesBefore = sortMap(frameBefore.resources);
+      auto sortedResourcesAfter = sortMap(frameAfter.resources);
+      auto sortedActionsBefore = sortMap(frameBefore.actions);
+      auto sortedActionsAfter = sortMap(frameAfter.actions);
+      auto sortedUnitsBefore = sortMap(frameBefore.units);
+      auto sortedUnitsAfter = sortMap(frameAfter.units);
+
+      // Lest tries to fancy-print EXPECT arguments.
+      // But it but doesn't know how to print these complex objects.
+      // So we do the comparison here and feed Lest the results.
+      auto matchingResources = sortedResourcesBefore == sortedResourcesAfter;
+      auto matchingActions = sortedActionsBefore == sortedActionsAfter;
+      auto matchingUnits = sortedUnitsBefore == sortedUnitsAfter;
```

Comment:
You should be able to compare `unordered_map` objects as they are, no?

Reply:
All of these comparisons confuse EXCEPT.